Pose2 measurement assertion error, Manifold

https://github.com/JuliaRobotics/RoME.jl/blob/bb2fe5f2cb8b9afae56bc742828311f144023642/src/factors/Pose2D.jl#L41

Oops, why a vector of `ProductRepr` for `Pose2`?
```julia
AssertionError: Pose2Pose2 expects measurement sample X to be a Manifolds tangent vector, not coordinate or point representation.  
Got X=
Manifolds.ProductRepr{
    Tuple{
        Vector{Float64}, 
        Matrix{Float64}}}[
            Manifolds.ProductRepr{
                Tuple{
                    Vector{Float64}, 
                    Matrix{Float64}}}(
                       ([339.55239021484266, -32.76530219142824], 
                        [0.0 0.1674510331200314; -0.1674510331200314 0.0])
                    )
            ]
```

Full stack

```julia
Task (failed) @0x00007f18399acd30
AssertionError: Pose2Pose2 expects measurement sample X to be a Manifolds tangent vector, not coordinate or point representation.  Got X=Manifolds.ProductRepr{Tuple{Vector{Float64}, Matrix{Float64}}}[Manifolds.ProductRepr{Tuple{Vector{Float64}, Matrix{Float64}}}(([339.55239021484266, -32.76530219142824], [0.0 0.1674510331200314; -0.1674510331200314 0.0]))]
Stacktrace:
  [1] (::CalcFactor{Pose2Pose2{FullNormal}, FactorMetadata{SubArray{DFGVariable{Pose2}, 1, Vector{DFGVariable{Pose2}}, Tuple{Vector{Int64}}, false}, SubArray{Symbol, 1, Vector{Symbol}, Tuple{Vector{Int64}}, false}, Tuple{Vector{Any}, Vector{Any}}, Nothing}, Vector{Tuple{Vector{Manifolds.ProductRepr{Tuple{Vector{Float64}, Matrix{Float64}}}}}}, Tuple{Vector{Any}, Vector{Any}}})(X::Vector{Manifolds.ProductRepr{Tuple{Vector{Float64}, Matrix{Float64}}}}, p::Manifolds.ProductRepr{Tuple{Vector{Float64}, Matrix{Float64}}}, q::Manifolds.ProductRepr{Tuple{Vector{Float64}, Matrix{Float64}}})
    @ RoME ~/.julia/dev/RoME/src/factors/Pose2D.jl:41
  [2] (::IncrementalInference.var"#271#273"{Int64, Vector{Tuple{Vector{Manifolds.ProductRepr{Tuple{Vector{Float64}, Matrix{Float64}}}}}}, CalcFactor{Pose2Pose2{FullNormal}, FactorMetadata{SubArray{DFGVariable{Pose2}, 1, Vector{DFGVariable{Pose2}}, Tuple{Vector{Int64}}, false}, SubArray{Symbol, 1, Vector{Symbol}, Tuple{Vector{Int64}}, false}, Tuple{Vector{Any}, Vector{Any}}, Nothing}, Vector{Tuple{Vector{Manifolds.ProductRepr{Tuple{Vector{Float64}, Matrix{Float64}}}}}}, Tuple{Vector{Any}, Vector{Any}}}, Tuple{Vector{Any}, Vector{Any}}})()
    @ IncrementalInference ~/.julia/dev/IncrementalInference/src/NumericalCalculations.jl:222
  [3] (::IncrementalInference.var"#_hypoObj#279"{SubArray{Any, 0, Vector{Any}, Tuple{Int64}, true}, IncrementalInference.var"#271#273"{Int64, Vector{Tuple{Vector{Manifolds.ProductRepr{Tuple{Vector{Float64}, Matrix{Float64}}}}}}, CalcFactor{Pose2Pose2{FullNormal}, FactorMetadata{SubArray{DFGVariable{Pose2}, 1, Vector{DFGVariable{Pose2}}, Tuple{Vector{Int64}}, false}, SubArray{Symbol, 1, Vector{Symbol}, Tuple{Vector{Int64}}, false}, Tuple{Vector{Any}, Vector{Any}}, Nothing}, Vector{Tuple{Vector{Manifolds.ProductRepr{Tuple{Vector{Float64}, Matrix{Float64}}}}}}, Tuple{Vector{Any}, Vector{Any}}}, Tuple{Vector{Any}, Vector{Any}}}})(x::Manifolds.ProductRepr{Tuple{Vector{Float64}, Matrix{Float64}}})
    @ IncrementalInference ~/.julia/dev/IncrementalInference/src/NumericalCalculations.jl:332
  [4] (::IncrementalInference.var"#cost#264"{IncrementalInference.var"#_hypoObj#279"{SubArray{Any, 0, Vector{Any}, Tuple{Int64}, true}, IncrementalInference.var"#271#273"{Int64, Vector{Tuple{Vector{Manifolds.ProductRepr{Tuple{Vector{Float64}, Matrix{Float64}}}}}}, CalcFactor{Pose2Pose2{FullNormal}, FactorMetadata{SubArray{DFGVariable{Pose2}, 1, Vector{DFGVariable{Pose2}}, Tuple{Vector{Int64}}, false}, SubArray{Symbol, 1, Vector{Symbol}, Tuple{Vector{Int64}}, false}, Tuple{Vector{Any}, Vector{Any}}, Nothing}, Vector{Tuple{Vector{Manifolds.ProductRepr{Tuple{Vector{Float64}, Matrix{Float64}}}}}}, Tuple{Vector{Any}, Vector{Any}}}, Tuple{Vector{Any}, Vector{Any}}}}, Manifolds.ProductRepr{Tuple{Vector{Float64}, Matrix{Float64}}}, Manifolds.GroupManifold{ℝ, Manifolds.ProductManifold{ℝ, Tuple{Manifolds.TranslationGroup{Tuple{2}, ℝ}, Manifolds.SpecialOrthogonal{2}}}, Manifolds.SemidirectProductOperation{Manifolds.RotationAction{Manifolds.TranslationGroup{Tuple{2}, ℝ}, Manifolds.SpecialOrthogonal{2}, Manifolds.LeftAction}}}})(p::Manifolds.ProductRepr{Tuple{Vector{Float64}, Matrix{Float64}}}, X::Manifolds.ProductRepr{Tuple{Vector{Float64}, Matrix{Float64}}}, Xc::Vector{Float64})
    @ IncrementalInference ~/.julia/dev/IncrementalInference/src/NumericalCalculations.jl:103
  [5] (::IncrementalInference.var"#263#265"{Manifolds.ProductRepr{Tuple{Vector{Float64}, Matrix{Float64}}}, Manifolds.ProductRepr{Tuple{Vector{Float64}, Matrix{Float64}}}, IncrementalInference.var"#cost#264"{IncrementalInference.var"#_hypoObj#279"{SubArray{Any, 0, Vector{Any}, Tuple{Int64}, true}, IncrementalInference.var"#271#273"{Int64, Vector{Tuple{Vector{Manifolds.ProductRepr{Tuple{Vector{Float64}, Matrix{Float64}}}}}}, CalcFactor{Pose2Pose2{FullNormal}, FactorMetadata{SubArray{DFGVariable{Pose2}, 1, Vector{DFGVariable{Pose2}}, Tuple{Vector{Int64}}, false}, SubArray{Symbol, 1, Vector{Symbol}, Tuple{Vector{Int64}}, false}, Tuple{Vector{Any}, Vector{Any}}, Nothing}, Vector{Tuple{Vector{Manifolds.ProductRepr{Tuple{Vector{Float64}, Matrix{Float64}}}}}}, Tuple{Vector{Any}, Vector{Any}}}, Tuple{Vector{Any}, Vector{Any}}}}, Manifolds.ProductRepr{Tuple{Vector{Float64}, Matrix{Float64}}}, Manifolds.GroupManifold{ℝ, Manifolds.ProductManifold{ℝ, Tuple{Manifolds.TranslationGroup{Tuple{2}, ℝ}, Manifolds.SpecialOrthogonal{2}}}, Manifolds.SemidirectProductOperation{Manifolds.RotationAction{Manifolds.TranslationGroup{Tuple{2}, ℝ}, Manifolds.SpecialOrthogonal{2}, Manifolds.LeftAction}}}}})(Xc::Vector{Float64})
    @ IncrementalInference ~/.julia/dev/IncrementalInference/src/NumericalCalculations.jl:116
  [6] value!!(obj::NLSolversBase.NonDifferentiable{Float64, Vector{Float64}}, x::Vector{Float64})
    @ NLSolversBase ~/.julia/packages/NLSolversBase/GRQ1x/src/interface.jl:9
  [7] initial_state(method::Optim.NelderMead{Optim.AffineSimplexer, Optim.AdaptiveParameters}, options::Optim.Options{Float64, Nothing}, d::NLSolversBase.NonDifferentiable{Float64, Vector{Float64}}, initial_x::Vector{Float64})
    @ Optim ~/.julia/packages/Optim/3K7JI/src/multivariate/solvers/zeroth_order/nelder_mead.jl:171
  [8] optimize
    @ ~/.julia/packages/Optim/3K7JI/src/multivariate/optimize/optimize.jl:35 [inlined]
  [9] optimize(f::Function, initial_x::Vector{Float64}, method::Optim.NelderMead{Optim.AffineSimplexer, Optim.AdaptiveParameters}, options::Optim.Options{Float64, Nothing}; inplace::Bool, autodiff::Symbol)
    @ Optim ~/.julia/packages/Optim/3K7JI/src/multivariate/optimize/interface.jl:142
 [10] optimize(f::Function, initial_x::Vector{Float64}, method::Optim.NelderMead{Optim.AffineSimplexer, Optim.AdaptiveParameters}, options::Optim.Options{Float64, Nothing}) (repeats 2 times)
    @ Optim ~/.julia/packages/Optim/3K7JI/src/multivariate/optimize/interface.jl:141
 [11] _solveLambdaNumeric(fcttype::Pose2Pose2{FullNormal}, objResX::IncrementalInference.var"#_hypoObj#279"{SubArray{Any, 0, Vector{Any}, Tuple{Int64}, true}, IncrementalInference.var"#271#273"{Int64, Vector{Tuple{Vector{Manifolds.ProductRepr{Tuple{Vector{Float64}, Matrix{Float64}}}}}}, CalcFactor{Pose2Pose2{FullNormal}, FactorMetadata{SubArray{DFGVariable{Pose2}, 1, Vector{DFGVariable{Pose2}}, Tuple{Vector{Int64}}, false}, SubArray{Symbol, 1, Vector{Symbol}, Tuple{Vector{Int64}}, false}, Tuple{Vector{Any}, Vector{Any}}, Nothing}, Vector{Tuple{Vector{Manifolds.ProductRepr{Tuple{Vector{Float64}, Matrix{Float64}}}}}}, Tuple{Vector{Any}, Vector{Any}}}, Tuple{Vector{Any}, Vector{Any}}}}, residual::Vector{Float64}, u0::Manifolds.ProductRepr{Tuple{Vector{Float64}, Matrix{Float64}}}, variableType::Pose2, islen1::Bool)
    @ IncrementalInference ~/.julia/dev/IncrementalInference/src/NumericalCalculations.jl:116
 [12] _solveCCWNumeric!(ccwl::CommonConvWrapper{Pose2Pose2{FullNormal}, Nothing, Vector{Int64}, NamedTuple{(:x4, :x5), Tuple{Vector{Any}, Vector{Any}}}, Nothing, Tuple{Vector{Manifolds.ProductRepr{Tuple{Vector{Float64}, Matrix{Float64}}}}}}; perturb::Float64, testshuffle::Bool, _slack::Nothing)
    @ IncrementalInference ~/.julia/dev/IncrementalInference/src/NumericalCalculations.jl:342
 [13] approxConvOnElements!(ccwl::CommonConvWrapper{Pose2Pose2{FullNormal}, Nothing, Vector{Int64}, NamedTuple{(:x4, :x5), Tuple{Vector{Any}, Vector{Any}}}, Nothing, Tuple{Vector{Manifolds.ProductRepr{Tuple{Vector{Float64}, Matrix{Float64}}}}}}, elements::Vector{Int64}, ::Type{SingleThreaded}, _slack::Nothing)
    @ IncrementalInference ~/.julia/dev/IncrementalInference/src/services/EvalFactor.jl:40
 [14] approxConvOnElements!(ccwl::CommonConvWrapper{Pose2Pose2{FullNormal}, Nothing, Vector{Int64}, NamedTuple{(:x4, :x5), Tuple{Vector{Any}, Vector{Any}}}, Nothing, Tuple{Vector{Manifolds.ProductRepr{Tuple{Vector{Float64}, Matrix{Float64}}}}}}, elements::Vector{Int64}, _slack::Nothing)
    @ IncrementalInference ~/.julia/dev/IncrementalInference/src/services/EvalFactor.jl:51
 [15] computeAcrossHypothesis!(ccwl::CommonConvWrapper{Pose2Pose2{FullNormal}, Nothing, Vector{Int64}, NamedTuple{(:x4, :x5), Tuple{Vector{Any}, Vector{Any}}}, Nothing, Tuple{Vector{Manifolds.ProductRepr{Tuple{Vector{Float64}, Matrix{Float64}}}}}}, hyporecipe::NamedTuple{(:certainidx, :allelements, :activehypo, :mhidx), Tuple{UnitRange{Int64}, Vector{Any}, Vector{Any}, Vector{Int64}}}, sfidx::Int64, maxlen::Int64, mani::Manifolds.GroupManifold{ℝ, Manifolds.ProductManifold{ℝ, Tuple{Manifolds.TranslationGroup{Tuple{2}, ℝ}, Manifolds.SpecialOrthogonal{2}}}, Manifolds.SemidirectProductOperation{Manifolds.RotationAction{Manifolds.TranslationGroup{Tuple{2}, ℝ}, Manifolds.SpecialOrthogonal{2}, Manifolds.LeftAction}}}; spreadNH::Float64, inflateCycles::Int64, skipSolve::Bool, _slack::Nothing)
    @ IncrementalInference ~/.julia/dev/IncrementalInference/src/services/EvalFactor.jl:190
 [16] evalPotentialSpecific(Xi::Vector{DFGVariable{Pose2}}, ccwl::CommonConvWrapper{Pose2Pose2{FullNormal}, Nothing, Vector{Int64}, NamedTuple{(:x4, :x5), Tuple{Vector{Any}, Vector{Any}}}, Nothing, Tuple{Vector{Manifolds.ProductRepr{Tuple{Vector{Float64}, Matrix{Float64}}}}}}, solvefor::Symbol, T_::Type{Pose2Pose2{FullNormal}}, measurement::Vector{Tuple}; needFreshMeasurements::Bool, solveKey::Symbol, N::Int64, spreadNH::Float64, inflateCycles::Int64, dbg::Bool, skipSolve::Bool, _slack::Nothing)
    @ IncrementalInference ~/.julia/dev/IncrementalInference/src/services/EvalFactor.jl:337
 [17] #evalPotentialSpecific#309
    @ ~/.julia/dev/IncrementalInference/src/services/EvalFactor.jl:476 [inlined]
 [18] evalFactor(dfg::LightDFG{SolverParams, DFGVariable, DFGFactor}, fct::DFGFactor{CommonConvWrapper{Pose2Pose2{FullNormal}, Nothing, Vector{Int64}, NamedTuple{(:x4, :x5), Tuple{Vector{Any}, Vector{Any}}}, Nothing, Tuple{Vector{Manifolds.ProductRepr{Tuple{Vector{Float64}, Matrix{Float64}}}}}}, 2}, solvefor::Symbol, measurement::Vector{Tuple}; needFreshMeasurements::Bool, solveKey::Symbol, N::Int64, inflateCycles::Int64, dbg::Bool, skipSolve::Bool, _slack::Nothing)
    @ IncrementalInference ~/.julia/dev/IncrementalInference/src/services/EvalFactor.jl:514
 [19] approxConvBelief(dfg::LightDFG{SolverParams, DFGVariable, DFGFactor}, fc::DFGFactor{CommonConvWrapper{Pose2Pose2{FullNormal}, Nothing, Vector{Int64}, NamedTuple{(:x4, :x5), Tuple{Vector{Any}, Vector{Any}}}, Nothing, Tuple{Vector{Manifolds.ProductRepr{Tuple{Vector{Float64}, Matrix{Float64}}}}}}, 2}, target::Symbol, measurement::Vector{Tuple}; solveKey::Symbol, N::Int64, skipSolve::Bool)
    @ IncrementalInference ~/.julia/dev/IncrementalInference/src/services/ApproxConv.jl:17
 [20] #calcProposalBelief#319
    @ ~/.julia/dev/IncrementalInference/src/services/ApproxConv.jl:161 [inlined]
 [21] proposalbeliefs!(dfg::LightDFG{SolverParams, DFGVariable, DFGFactor}, destlbl::Symbol, factors::Vector{DFGFactor{CommonConvWrapper{Pose2Pose2{FullNormal}, Nothing, Vector{Int64}, NamedTuple{(:x4, :x5), Tuple{Vector{Any}, Vector{Any}}}, Nothing, Tuple{Vector{Manifolds.ProductRepr{Tuple{Vector{Float64}, Matrix{Float64}}}}}}, 2}}, dens::Vector{ManifoldKernelDensity}, measurement::Vector{Tuple}; solveKey::Symbol, N::Int64, dbg::Bool)
    @ IncrementalInference ~/.julia/dev/IncrementalInference/src/services/ApproxConv.jl:225
 [22] propagateBelief(dfg::LightDFG{SolverParams, DFGVariable, DFGFactor}, destvar::DFGVariable{Pose2}, factors::Vector{DFGFactor{CommonConvWrapper{Pose2Pose2{FullNormal}, Nothing, Vector{Int64}, NamedTuple{(:x4, :x5), Tuple{Vector{Any}, Vector{Any}}}, Nothing, Tuple{Vector{Manifolds.ProductRepr{Tuple{Vector{Float64}, Matrix{Float64}}}}}}, 2}}; solveKey::Symbol, dens::Vector{ManifoldKernelDensity}, N::Int64, needFreshMeasurements::Bool, dbg::Bool, logger::Base.CoreLogging.SimpleLogger)
    @ IncrementalInference ~/.julia/dev/IncrementalInference/src/GraphProductOperations.jl:30
 [23] doautoinit!(dfg::LightDFG{SolverParams, DFGVariable, DFGFactor}, xi::DFGVariable{Pose2}; solveKey::Symbol, singles::Bool, N::Int64, logger::Base.CoreLogging.SimpleLogger)
    @ IncrementalInference ~/.julia/dev/IncrementalInference/src/services/GraphInit.jl:125
 [24] #doautoinit!#16
    @ ~/.julia/dev/IncrementalInference/src/services/GraphInit.jl:159 [inlined]
 [25] cycleInitByVarOrder!(subfg::LightDFG{SolverParams, DFGVariable, DFGFactor}, varorder::Vector{Symbol}; solveKey::Symbol, logger::Base.CoreLogging.SimpleLogger)
    @ IncrementalInference ~/.julia/dev/IncrementalInference/src/SolveTree.jl:250
 [26] solveUp_StateMachine(csmc::CliqStateMachineContainer{BayesTreeNodeData, LightDFG{SolverParams, DFGVariable, DFGFactor}, LightDFG{SolverParams, DFGVariable, DFGFactor}, MetaBayesTree})
    @ IncrementalInference ~/.julia/dev/IncrementalInference/src/CliqueStateMachine.jl:358
 [27] (::StateMachine{CliqStateMachineContainer})(userdata::CliqStateMachineContainer{BayesTreeNodeData, LightDFG{SolverParams, DFGVariable, DFGFactor}, LightDFG{SolverParams, DFGVariable, DFGFactor}, MetaBayesTree}, timeout::Nothing; pollinterval::Float64, breakafter::Function, verbose::Bool, verbosefid::Base.TTY, verboseXtra::IncrementalInference.CliqStatus, iterlimit::Int64, injectDelayBefore::Nothing, recordhistory::Bool, housekeeping_cb::IncrementalInference.var"#469#471"{IncrementalInference.TreeClique, CliqStateMachineContainer{BayesTreeNodeData, LightDFG{SolverParams, DFGVariable, DFGFactor}, LightDFG{SolverParams, DFGVariable, DFGFactor}, MetaBayesTree}})
    @ FunctionalStateMachine ~/.julia/packages/FunctionalStateMachine/2JZFG/src/StateMachine.jl:82
 [28] initStartCliqStateMachine!(dfg::LightDFG{SolverParams, DFGVariable, DFGFactor}, tree::MetaBayesTree, cliq::IncrementalInference.TreeClique, timeout::Nothing; oldcliqdata::BayesTreeNodeData, verbose::Bool, verbosefid::Base.TTY, drawtree::Bool, show::Bool, incremental::Bool, limititers::Int64, upsolve::Bool, downsolve::Bool, recordhistory::Bool, delay::Bool, logger::Base.CoreLogging.SimpleLogger, solve_progressbar::Nothing, algorithm::Symbol, solveKey::Symbol)
    @ IncrementalInference ~/.julia/dev/IncrementalInference/src/CliqueStateMachine.jl:64
 [29] tryCliqStateMachineSolve!(dfg::LightDFG{SolverParams, DFGVariable, DFGFactor}, treel::MetaBayesTree, cliqKey::Int64, timeout::Nothing; oldtree::MetaBayesTree, verbose::Bool, verbosefid::Base.TTY, drawtree::Bool, limititers::Int64, downsolve::Bool, incremental::Bool, delaycliqs::Vector{Symbol}, recordcliqs::Vector{Symbol}, solve_progressbar::Nothing, algorithm::Symbol, solveKey::Symbol)
    @ IncrementalInference ~/.julia/dev/IncrementalInference/src/SolverAPI.jl:112
 [30] (::IncrementalInference.var"#522#525"{MetaBayesTree, Bool, Bool, Base.TTY, Bool, Bool, Vector{Symbol}, Vector{Symbol}, Symbol, Symbol, LightDFG{SolverParams, DFGVariable, DFGFactor}, MetaBayesTree, Nothing, Nothing, Int64, Int64})()
    @ IncrementalInference ./task.jl:411
```

--- a/src/factors/Pose2D.jl
+++ b/src/factors/Pose2D.jl
@@ -24,14 +24,13 @@ DFG.getManifold(::Pose2Pose2) = Manifolds.SpecialEuclidean(2)
 
 Pose2Pose2(::UniformScaling) = Pose2Pose2() # MvNormal(zeros(3),LinearAlgebra.diagm([1.0;1.0;1.0])) )
 
-function getSample(cf::CalcFactor{<:Pose2Pose2}, N::Int=1) 
+
+function getSample(cf::CalcFactor{<:Pose2Pose2}) 
   
-  Xc = [rand(cf.factor.z) for _ in 1:N]
-  #NOTE be careful using this as template as it will not work in general for all manifolds
-  M = getManifold(Pose2)
-  # ϵ = getPointIdentity(Pose2)
-  X = hat.(Ref(M), Ref(Manifolds.Identity(M)), Xc)
-  # return a vector
+  M = getManifold(cf.factor)
+  ϵ = getPointIdentity(Pose2)
+
+  X = sampleTangent(M, cf.factor.z, ϵ)
   return (X, )
 end
 

